### PR TITLE
ETH Wallet Age strategy

### PR DESCRIPTION
**Pull request by ChaituVR**
# ETH Wallet Age strategy

Allows getting the first transaction date from etherscan, which are used to calculate voter scores with a number of days from the first transaction.

If this is successful we can add ERC20 token age very similarly

### CODE REVIEW

File: src/strategies/eth-wallet-age/index.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import fetch from 'cross-fetch';
+
+export const author = 'chaituvr';
+export const version = '0.1.0';
+
+const delay = interval => new Promise(resolve => setTimeout(resolve, interval));
```
Comment (bonustrack) [inline src/strategies/eth-wallet-age/index.ts:6]: This doesnt seem to be used

File: src/strategies/eth-wallet-age/index.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+export async function strategy(
+  space,
+  network,
+  provider,
+  addresses,
+  options,
+  snapshot
+  ) {
+  const apiLimit = 300;
```
Comment (bonustrack) [inline src/strategies/eth-wallet-age/index.ts:16]: Why you need this?

Comment (ChaituVR) [reply src/strategies/eth-wallet-age/index.ts:16]: Etherscan has an API limit - https://info.etherscan.com/api-return-errors/

```5 calls per sec/IP with a valid API key```

so 1000/5 ms should be the gap between each API call.. even if we give 200 sometimes it is exceeding the limit. works great with 300

File: src/strategies/eth-wallet-age/index.ts (hunk 1 above)
Comment (ChaituVR) [reply src/strategies/eth-wallet-age/index.ts:6]: Fixed It 👍

File: src/strategies/eth-wallet-age/index.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  let data = []
+
+  data = await Promise.all(addresses.map((address, index) => {
+    return new Promise(resolve => setTimeout(resolve, apiLimit * index))
```
Comment (bonustrack) [inline src/strategies/eth-wallet-age/index.ts:20]: We can't send a request for each voter, some proposal have 1000 votes, we dont want the user browser to send 1000 requests. Not sure how you can bypass this.

Comment (bonustrack) [reply src/strategies/eth-wallet-age/index.ts:20]: Maybe using an API like the one used here can work https://github.com/snapshot-labs/snapshot.js/pull/60

**Comment by ChaituVR** [comment]
@bonustrack Replace dfuse API instead of etherscan

File: src/strategies/eth-wallet-age/examples.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
+[
+  {
+    "name": "Age in Days",
+    "strategy": {
+      "name": "eth-wallet-age",
+      "params": {
+        "dfuseApiKey": "web_f527db575a38dd11c5b686d7da54d371"
```
Comment (bonustrack) [inline src/strategies/eth-wallet-age/examples.json:7]: I think this API Key should be hardcoded within the strategy. Is there some limitation with the API?

File: src/strategies/eth-wallet-age/index.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      },
+      node: {
+        from: true,
+        to: true
+      }
+    }
+  }]));
+  
+  const dfuseJWT = await getJWT(options.dfuseApiKey || 'web_f527db575a38dd11c5b686d7da54d371');
+  data = await subgraphRequest('https://mainnet.eth.dfuse.io/graphql', query, {
```
Comment (bonustrack) [inline src/strategies/eth-wallet-age/index.ts:54]: Is this available on others network? Rinkeby, Kovan etc? If yes we may want to handle it with something like this: https://github.com/snapshot-labs/snapshot.js/blob/master/src/strategies/balancer/index.ts#L7-L10

File: src/strategies/eth-wallet-age/index.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    headers: {
+      Authorization: 'Bearer ' + dfuseJWT
```
Comment (bonustrack) [inline src/strategies/eth-wallet-age/index.ts:56]: We should always use this way
```suggestion
      Authorization: `Bearer ${dfuseJWT}`
```

File: src/strategies/eth-wallet-age/index.ts (hunk 5 above)
Comment (ChaituVR) [reply src/strategies/eth-wallet-age/index.ts:54]: Only available for Mainnet and Ropsten... but Ropsten isn't available in https://github.com/snapshot-labs/snapshot.js/blob/master/src/networks.json  🤔

File: src/strategies/eth-wallet-age/index.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
+import { EnumType } from 'json-to-graphql-query';
+import fetch from 'cross-fetch';
+import { subgraphRequest } from '../../utils';
+
+export const author = 'chaituvr';
```
Comment (bonustrack) [inline src/strategies/eth-wallet-age/index.ts:5]:
```suggestion
export const author = 'ChaituVR';
```

**Review by bonustrack — APPROVED** [review]
Well done!

**Pull request merged**